### PR TITLE
Unroll calls to any #5062

### LINKED ISSUE #5062: Unroll the iterable for all/any calls to get better reports
Sometime I need to assert some predicate on all of an iterable, and for that the builtin functions `all`/`any` are great - but the failure messages aren't useful at all!
For example - the same test written in three ways:

- A generator expression
```sh                                                                                                                                                                                                                         
    def test_all_even():
        even_stevens = list(range(1,100,2))
>       assert all(is_even(number) for number in even_stevens)
E       assert False
E        +  where False = all(<generator object test_all_even.<locals>.<genexpr> at 0x101f82ed0>)
```
- A list comprehension
```sh
    def test_all_even():
        even_stevens = list(range(1,100,2))
>       assert all([is_even(number) for number in even_stevens])
E       assert False
E        +  where False = all([False, False, False, False, False, False, ...])
```
- A for loop
```sh
    def test_all_even():
        even_stevens = list(range(1,100,2))
        for number in even_stevens:
>           assert is_even(number)
E           assert False
E            +  where False = is_even(1)

test_all_any.py:7: AssertionError
```
The only one that gives a meaningful report is the for loop - but it's way more wordy, and `all` asserts don't translate to a for loop nicely (I'll have to write a `break` or a helper function - yuck)
I propose the assertion re-writer "unrolls" the iterator to the third form, and then uses the already existing reports.

- [x] Include a detailed description of the bug or suggestion
- [x] `pip list` of the virtual environment you are using
```
Package        Version
-------------- -------
atomicwrites   1.3.0  
attrs          19.1.0 
more-itertools 7.0.0  
pip            19.0.3 
pluggy         0.9.0  
py             1.8.0  
pytest         4.4.0  
setuptools     40.8.0 
six            1.12.0 
```
- [x] pytest and operating system versions
`platform darwin -- Python 3.7.3, pytest-4.4.0, py-1.8.0, pluggy-0.9.0`
- [x] Minimal example if possible

### PR DESCRIPTION
**Pull request by Tadaboody**
Closes #5062 

- [x] Create a new changelog file in the `changelog` folder, with a name like `<ISSUE NUMBER>.<TYPE>.rst`. See [changelog/README.rst](https://github.com/pytest-dev/pytest/blob/master/changelog/README.rst) for details.
- [x] Target the `master` branch for bug fixes, documentation updates and trivial changes.
- [x] Target the `features` branch for new features and removals/deprecations.
- [ ] Include documentation when adding new features.
- [x] Include new tests or update existing tests when applicable.

### CODE REVIEW

**Comment by Tadaboody** [comment]
Current roadblock - I want to recursively rewrite the assert inside the for loop, but `visit_Assert` resets the state so I can't call `self.visit(assert_node)`. On the other hand, I want to stay consistent with things like variable names so I don't want to create a new `AssertionVisitor`

**Comment by Tadaboody** [comment]
Ok it's at a point where it shows a better report but not quite detailed enough, would like a proposition on where to take it from here
```py
test_example.py F                                                        [100%]

=================================== FAILURES ===================================
_________________________________ test_rolled __________________________________

    def test_rolled():
>       assert all(x == 1 for x in range(10))
E       assert 0 == 1

test_example.py:2: AssertionError

```
The rewritten ast is exactly as expected, just as if it was a for loop:
```
=========================== Rewritten AST as Python ============================
import builtins as @py_builtins
import _pytest.assertion.rewrite as @pytest_ar

def test_rolled():
    for x in range(10):
        import builtins as @py_builtins
        import _pytest.assertion.rewrite as @pytest_ar
        @py_assert2 = 1
        @py_assert1 = x == @py_assert2
        if not @py_assert1:
            @py_format4 = @pytest_ar._call_reprcompare(('==',), (@py_assert1,), ('%(py0)s == %(py3)s',), (x, @py_assert2)) % {'py0': @pytest_ar._saferepr(x) if 'x' in @py_builtins.locals() or @pytest_ar._should_repr_global_name(x) else 'x', 'py3': @pytest_ar._saferepr(@py_assert2)}
            @py_format6 = ('' + 'assert %(py5)s') % {'py5': @py_format4}
            raise AssertionError(@pytest_ar._format_explanation(@py_format6))
        @py_assert1 = @py_assert2 = None
    if 1 is None:
        from _pytest.warning_types import PytestWarning
        from warnings import warn_explicit
        warn_explicit(PytestWarning('asserting the value None, please use "assert is None"'), category=None, filename='/Users/me/Forks/pytest/test_example.py', lineno=2)
    if not 1:
        @py_format0 = ('' + 'assert ') % {}
        raise AssertionError(@pytest_ar._format_explanation(@py_format0))
=========================== short test summary info ============================
FAILED test_example.py::test_rolled
=========================== 1 failed in 0.10 seconds ===========================
```

**Comment by Tadaboody** [comment]
I'm not going to be able to continue this until next week - @danielx123 if you want to fork this until then you're more than welcome!

File: src/_pytest/assertion/rewrite.py [hunk 1]
```diff
@@ -987,6 +989,27 @@ def visit_Call_35(self, call):
         outer_expl = "%s\n{%s = %s\n}" % (res_expl, res_expl, expl)
         return res, outer_expl
 
+    def visit_all(self, call):
+        """Special rewrite for the builtin all function, see #5602"""
```
Comment (Stranger6667) [inline src/_pytest/assertion/rewrite.py]: Seems like it should be #5062 instead of #5602

**Comment by danielx123** [comment]
@Tadaboody I can work on this during the weekend. How do you want to work on this together?

**Comment by Tadaboody** [comment]
Alright I think this is pretty stable and does what it's supposed to @nicoddemus I'll be waiting to start a review process

**Comment by nicoddemus** [comment]
Do you mind rebasing on `features` and pushing again? `codecov` is providing some funny results.

**Comment by Tadaboody** [comment]
rebased

**Comment by nicoddemus** [comment]
Merging as the failure is unrelated: #5317

**Comment by nicoddemus** [comment]
Thanks @Tadaboody!

**Comment by blueyed** [comment]
@nicoddemus 

> Merging as the failure is unrelated: #5317

It turns out that it is not really it seems.. ;)

**Comment by nicoddemus** [comment]
😮

**Pull request merged**